File: worker/src/acme.ts
import type { CertificateResult, Env } from './types'
import {
  generateRsaKeyPair,
  importPrivateKeyFromPem,
  keyAuthorizationDigest,
  privateKeyToPem,
  publicJwkFromPrivateKey,
  signRs256,
} from './crypto'
import { createCsr, csrToAcmeBase64, parseCertificateExpiresAt } from './asn1'
import { getAcmeAccount, saveAcmeAccount, saveEvent } from './db'
import { base64url, nowIso } from './utils'

const ACME_PROD_DIRECTORY = 'https://acme-v02.api.letsencrypt.org/directory'
const ACME_STAGING_DIRECTORY = 'https://acme-staging-v02.api.letsencrypt.org/directory'
const CF_API_BASE = 'https://api.cloudflare.com/client/v4'

interface AcmeDirectory {
  newNonce: string
  newAccount: string
  newOrder: string
}

interface AcmeOrder {
  status: string
  authorizations: string[]
  finalize: string
  certificate?: string
  _url: string
}

interface AcmeAuthorization {
  status: string
  identifier: { type: string, value: string }
  challenges: Array<{ type: string, token: string, url: string, status?: string }>
}

interface CloudflareResponse<T> {
  success: boolean
  result: T
  errors?: unknown[]
}

interface AcmeClientState {
  directory: AcmeDirectory
  nonce: string
  accountKey: CryptoKey
  accountJwk: JsonWebKey
  accountUrl: string | null
}

interface ApplyRuntime {
  sleepDns(): Promise<void>
  sleepPoll(name: string): Promise<void>
}

type LogLevel = 'debug' | 'info' | 'success' | 'warn' | 'error'

async function log(env: Env, domain: string, message: string, level: LogLevel = 'info', step = ''): Promise<void> {
  await saveEvent(env, domain, step, level, message)
}

async function fetchJson<T>(url: string, init?: RequestInit): Promise<{ data: T, response: Response }> {
  const response = await fetch(url, init)
  const text = await response.text()
  let data: T
  try {
    data = text ? JSON.parse(text) as T : {} as T
  } catch {
    throw new Error(`HTTP ${response.status}: ${text || 'empty response'}`)
  }
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${text}`)
  }
  return { data, response }
}

async function getDirectory(staging: boolean): Promise<AcmeDirectory> {
  const { data } = await fetchJson<AcmeDirectory>(staging ? ACME_STAGING_DIRECTORY : ACME_PROD_DIRECTORY)
  return data
}

async function getNonce(directory: AcmeDirectory): Promise<string> {
  const response = await fetch(directory.newNonce, { method: 'HEAD' })
  const nonce = response.headers.get('Replay-Nonce')
  if (!nonce) throw new Error('ACME server did not return a Replay-Nonce')
  return nonce
}

async function loadOrCreateAccountKey(env: Env, domain: string): Promise<{
  key: CryptoKey
  jwk: JsonWebKey
  keyPem: string
  accountUrl: string | null
}> {
  const account = await getAcmeAccount(env, domain)
  if (account?.account_key_pem) {
    const key = await importPrivateKeyFromPem(account.account_key_pem)
    return {
      key,
      jwk: await publicJwkFromPrivateKey(key),
      keyPem: account.account_key_pem,
      accountUrl: account.account_url,
    }
  }

  const keyPair = await generateRsaKeyPair()
  const keyPem = await privateKeyToPem(keyPair.privateKey)
  return {
    key: keyPair.privateKey,
    jwk: await publicJwkFromPrivateKey(keyPair.privateKey),
    keyPem,
    accountUrl: null,
  }
}

async function createAcmeState(env: Env, domain: string, staging: boolean): Promise<AcmeClientState & { keyPem: string }> {
  const directory = await getDirectory(staging)
  const nonce = await getNonce(directory)
  const account = await loadOrCreateAccountKey(env, domain)
  return {
    directory,
    nonce,
    accountKey: account.key,
    accountJwk: account.jwk,
    accountUrl: account.accountUrl,
    keyPem: account.keyPem,
  }
}

async function signJws(state: AcmeClientState, url: string, payload: unknown): Promise<Record<string, string>> {
  const header: Record<string, unknown> = {
    alg: 'RS256',
    nonce: state.nonce,
    url,
  }
  if (state.accountUrl) {
    header.kid = state.accountUrl
  } else {
    header.jwk = state.accountJwk
  }

  const protected64 = base64url(JSON.stringify(header))
  const payload64 = payload === null ? '' : base64url(JSON.stringify(payload))
  const signature = await signRs256(state.accountKey, `${protected64}.${payload64}`)
  return {
    protected: protected64,
    payload: payload64,
    signature: base64url(signature),
  }
}

async function acmePost<T>(state: AcmeClientState, url: string, payload: unknown): Promise<{ data: T, response: Response }> {
  const body = await signJws(state, url, payload)
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/jose+json' },
    body: JSON.stringify(body),
  })
  const nonce = response.headers.get('Replay-Nonce')
  if (nonce) state.nonce = nonce
  const text = await response.text()
  let data: T
  try {
    data = text ? JSON.parse(text) as T : {} as T
  } catch {
    data = text as T
  }
  if (!response.ok) {
    throw new Error(`ACME HTTP ${response.status}: ${text}`)
  }
  return { data, response }
}

async function acmePostAsText(state: AcmeClientState, url: string, payload: unknown): Promise<string> {
  const body = await signJws(state, url, payload)
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/jose+json' },
    body: JSON.stringify(body),
  })
  const nonce = response.headers.get('Replay-Nonce')
  if (nonce) state.nonce = nonce
  const text = await response.text()
  if (!response.ok) {
    throw new Error(`ACME HTTP ${response.status}: ${text}`)
  }
  return text
}

async function ensureAccount(env: Env, domain: string, state: AcmeClientState & { keyPem: string }): Promise<void> {
  if (state.accountUrl) return
  const { response } = await acmePost(state, state.directory.newAccount, { termsOfServiceAgreed: true })
  state.accountUrl = response.headers.get('Location')
  if (!state.accountUrl) {
    throw new Error('ACME account registration did not return Location header')
  }
  await saveAcmeAccount(env, domain, state.keyPem, state.accountUrl)
}

async function createOrder(state: AcmeClientState, domains: string[]): Promise<AcmeOrder> {
  const identifiers = domains.map(value => ({ type: 'dns', value }))
  const { data, response } = await acmePost<Omit<AcmeOrder, '_url'>>(state, state.directory.newOrder, { identifiers })
  const orderUrl = response.headers.get('Location')
  if (!orderUrl) throw new Error('ACME order did not return Location header')
  return { ...data, _url: orderUrl }
}

async function getAuthorization(state: AcmeClientState, url: string): Promise<AcmeAuthorization> {
  const { data } = await acmePost<AcmeAuthorization>(state, url, null)
  return data
}

async function respondChallenge(state: AcmeClientState, url: string): Promise<void> {
  await acmePost(state, url, {})
}

async function pollStatus<T extends { status?: string }>(
  state: AcmeClientState,
  url: string,
  target: string,
  runtime: ApplyRuntime,
  label: string,
  maxTries = 30,
): Promise<T> {
  for (let i = 0; i < maxTries; i++) {
    const { data } = await acmePost<T>(state, url, null)
    if (data.status === target) return data
    if (data.status && ['invalid', 'expired', 'revoked'].includes(data.status)) {
      throw new Error(`状态异常: ${data.status} - ${JSON.stringify(data)}`)
    }
    await runtime.sleepPoll(`${label}-${i + 1}`)
  }
  throw new Error('轮询超时')
}

async function finalizeOrder(state: AcmeClientState, url: string, csrDer: Uint8Array): Promise<AcmeOrder> {
  const { data } = await acmePost<AcmeOrder>(state, url, { csr: csrToAcmeBase64(csrDer) })
  return data
}

async function downloadCertificate(state: AcmeClientState, url: string): Promise<string> {
  return acmePostAsText(state, url, null)
}

function cfHeaders(token: string): HeadersInit {
  return {
    Authorization: `Bearer ${token}`,
    'Content-Type': 'application/json',
  }
}

export async function verifyCloudflareToken(token: string): Promise<boolean> {
  const { data } = await fetchJson<CloudflareResponse<unknown>>(`${CF_API_BASE}/user/tokens/verify`, {
    headers: cfHeaders(token),
  })
  return data.success
}

async function getZoneId(domain: string, token: string): Promise<string> {
  const url = new URL(`${CF_API_BASE}/zones`)
  url.searchParams.set('name', domain)
  const { data } = await fetchJson<CloudflareResponse<Array<{ id: string }>>>(url.toString(), {
    headers: cfHeaders(token),
  })
  if (!data.success || !data.result?.length) {
    throw new Error(`未找到域名 ${domain} 的 Zone，请确认域名已添加到 Cloudflare`)
  }
  return data.result[0].id
}

async function createTxtRecord(zoneId: string, name: string, content: string, token: string): Promise<string> {
  const { data } = await fetchJson<CloudflareResponse<{ id: string }>>(`${CF_API_BASE}/zones/${zoneId}/dns_records`, {
    method: 'POST',
    headers: cfHeaders(token),
    body: JSON.stringify({ type: 'TXT', name, content, ttl: 120 }),
  })
  if (!data.success) {
    throw new Error(`创建 DNS 记录失败: ${JSON.stringify(data.errors || data)}`)
  }
  return data.result.id
}

async function deleteTxtRecord(zoneId: string, recordId: string, token: string): Promise<void> {
  await fetch(`${CF_API_BASE}/zones/${zoneId}/dns_records/${recordId}`, {
    method: 'DELETE',
    headers: cfHeaders(token),
  })
}

export async function applyCertificate(
  env: Env,
  domain: string,
  cfToken: string,
  staging: boolean,
  runtime: ApplyRuntime,
): Promise<CertificateResult> {
  const domains = [domain, `*.${domain}`]
  const dnsRecords: Array<{ zoneId: string, recordId: string }> = []

  try {
    await log(env, domain, `开始申请证书: ${domains.join(', ')}`, 'info', 'init')

    await log(env, domain, "连接 Let's Encrypt...", 'info', 'init')
    const state = await createAcmeState(env, domain, staging)
    await log(env, domain, '连接成功', 'success', 'init')

    await log(env, domain, '准备 ACME 账户...', 'info', 'account')
    await ensureAccount(env, domain, state)
    await log(env, domain, '账户就绪', 'success', 'account')

    await log(env, domain, '创建证书订单...', 'info', 'order')
    const order = await createOrder(state, domains)
    await log(env, domain, `订单已创建，需验证 ${order.authorizations.length} 个授权`, 'success', 'order')

    await log(env, domain, `查询 Cloudflare Zone: ${domain}`, 'info', 'dns')
    const zoneId = await getZoneId(domain, cfToken)
    await log(env, domain, 'Zone 已找到', 'success', 'dns')

    const challenges: Array<{ authUrl: string, challengeUrl: string, authDomain: string }> = []
    for (const authUrl of order.authorizations) {
      const auth = await getAuthorization(state, authUrl)
      const authDomain = auth.identifier.value.replace(/^\*\./, '')
      const challenge = auth.challenges.find(item => item.type === 'dns-01')
      if (!challenge) {
        throw new Error(`未找到 ${authDomain} 的 dns-01 验证`)
      }

      const txtValue = await keyAuthorizationDigest(challenge.token, state.accountJwk)
      const recordName = `_acme-challenge.${authDomain}`
      await log(env, domain, `添加 DNS 记录: ${recordName}`, 'info', 'dns')
      const recordId = await createTxtRecord(zoneId, recordName, txtValue, cfToken)
      dnsRecords.push({ zoneId, recordId })
      challenges.push({ authUrl, challengeUrl: challenge.url, authDomain })
      await log(env, domain, 'DNS 记录已添加', 'success', 'dns')
    }

    await log(env, domain, '等待 DNS 传播...', 'info', 'propagation')
    await runtime.sleepDns()
    await log(env, domain, 'DNS 传播完成', 'success', 'propagation')

    for (const challenge of challenges) {
      await log(env, domain, `提交验证: ${challenge.authDomain}`, 'info', 'verify')
      await respondChallenge(state, challenge.challengeUrl)
    }

    for (const challenge of challenges) {
      await log(env, domain, `等待验证: ${challenge.authDomain}...`, 'info', 'verify')
      await pollStatus<AcmeAuthorization>(
        state,
        challenge.authUrl,
        'valid',
        runtime,
        `poll-auth-${challenge.authDomain}`,
      )
      await log(env, domain, `${challenge.authDomain} 验证通过`, 'success', 'verify')
    }

    await log(env, domain, '生成证书密钥和 CSR...', 'info', 'generate')
    const { csrDer, keyPem } = await createCsr(domains)
    await log(env, domain, 'CSR 已生成', 'success', 'generate')

    await log(env, domain, '提交订单完成请求...', 'info', 'finalize')
    let result = await finalizeOrder(state, order.finalize, csrDer)
    if (result.status !== 'valid') {
      result = await pollStatus<AcmeOrder>(state, order._url, 'valid', runtime, 'poll-order')
    }
    if (!result.certificate) {
      throw new Error('ACME 订单没有返回证书下载地址')
    }

    await log(env, domain, '下载证书...', 'info', 'finalize')
    const fullchainPem = await downloadCertificate(state, result.certificate)
    await log(env, domain, '证书已下载', 'success', 'finalize')

    const issuedAt = nowIso()
    const expiresAt = parseCertificateExpiresAt(fullchainPem)
      || new Date(Date.now() + 90 * 24 * 60 * 60 * 1000).toISOString()
    const metadata = {
      domain,
      domains,
      issued_at: issuedAt,
      expires_at: expiresAt,
      acme_directory: staging ? ACME_STAGING_DIRECTORY : ACME_PROD_DIRECTORY,
    }

    return {
      fullchainPem,
      privkeyPem: keyPem,
      metadata,
      issuedAt,
      expiresAt,
    }
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error)
    await log(env, domain, `申请失败: ${message}`, 'error', 'error')
    throw error
  } finally {
    if (dnsRecords.length) {
      await log(env, domain, '清理 DNS 记录...', 'info', 'cleanup')
      for (const record of dnsRecords) {
        try {
          await deleteTxtRecord(record.zoneId, record.recordId, cfToken)
        } catch (error) {
          const message = error instanceof Error ? error.message : String(error)
          await log(env, domain, `清理失败: ${message}`, 'warn', 'cleanup')
        }
      }
      await log(env, domain, 'DNS 记录已清理', 'success', 'cleanup')
    }
  }
}

File: backend/app/services/acme.py
"""ACME v2 协议服务 - 纯 Python 实现 Let's Encrypt 证书申请"""

import asyncio
import base64
import hashlib
import json
from datetime import datetime, timezone
from typing import Any, Callable, List, Optional, Tuple

import httpx
from cryptography import x509
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15
from cryptography.x509.oid import NameOID

from app.db import database as db
from app.services.cloudflare import CloudflareService

LETS_ENCRYPT_DIR = "https://acme-v02.api.letsencrypt.org/directory"
LETS_ENCRYPT_STAGING_DIR = "https://acme-staging-v02.api.letsencrypt.org/directory"


def _b64url(data) -> str:
    """Base64url 编码（无 padding）"""
    if isinstance(data, str):
        data = data.encode()
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def _int_to_b64url(n: int) -> str:
    """将大整数转为 base64url 编码"""
    byte_len = (n.bit_length() + 7) // 8
    return _b64url(n.to_bytes(byte_len, byteorder="big"))


class AcmeService:
    """ACME v2 客户端，支持 DNS-01 验证申请通配符证书"""

    def __init__(self, staging: bool = False):
        self.directory_url = LETS_ENCRYPT_STAGING_DIR if staging else LETS_ENCRYPT_DIR
        self.directory: Optional[dict] = None
        self.account_key = None
        self.account_key_pem: bytes | None = None
        self.account_url: Optional[str] = None
        self.nonce: Optional[str] = None
        self.client = httpx.AsyncClient(timeout=60.0)
        self._log_cb: Optional[Callable] = None

    def set_log_callback(self, cb: Callable):
        self._log_cb = cb

    async def _log(self, msg: str, level: str = "info", step: str = ""):
        if self._log_cb:
            await self._log_cb(msg, level, step)

    # ── JWK / JWS ─────────────────────────────────────────────

    def _get_jwk(self) -> dict:
        pub = self.account_key.public_key().public_numbers()
        return {"e": _int_to_b64url(pub.e), "kty": "RSA", "n": _int_to_b64url(pub.n)}

    def _thumbprint(self) -> str:
        jwk = self._get_jwk()
        ordered = json.dumps(jwk, sort_keys=True, separators=(",", ":"))
        return _b64url(hashlib.sha256(ordered.encode()).digest())

    def _sign_jws(self, url: str, payload) -> dict:
        header = {"alg": "RS256", "nonce": self.nonce, "url": url}
        if self.account_url:
            header["kid"] = self.account_url
        else:
            header["jwk"] = self._get_jwk()

        protected_b64 = _b64url(json.dumps(header))
        payload_b64 = "" if payload is None else _b64url(json.dumps(payload))

        sig = self.account_key.sign(
            f"{protected_b64}.{payload_b64}".encode(), PKCS1v15(), hashes.SHA256()
        )
        return {"protected": protected_b64, "payload": payload_b64, "signature": _b64url(sig)}

    async def _post(self, url: str, payload=None) -> httpx.Response:
        body = self._sign_jws(url, payload)
        resp = await self.client.post(
            url, json=body, headers={"Content-Type": "application/jose+json"}
        )
        if "Replay-Nonce" in resp.headers:
            self.nonce = resp.headers["Replay-Nonce"]
        return resp

    # ── ACME 流程 ──────────────────────────────────────────────

    async def init(self, domain: str):
        """初始化：获取目录、Nonce、加载/生成账户密钥"""
        resp = await self.client.get(self.directory_url)
        self.directory = resp.json()

        resp = await self.client.head(self.directory["newNonce"])
        self.nonce = resp.headers["Replay-Nonce"]

        account = await db.get_acme_account(domain)
        if account and account.get("account_key_pem"):
            self.account_key_pem = account["account_key_pem"].encode("utf-8")
            self.account_key = serialization.load_pem_private_key(self.account_key_pem, password=None)
            self.account_url = account.get("account_url")
        else:
            self.account_key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
            self.account_key_pem = self.account_key.private_bytes(
                serialization.Encoding.PEM,
                serialization.PrivateFormat.PKCS8,
                serialization.NoEncryption(),
            )

    async def ensure_account(self, domain: str):
        if self.account_url:
            return
        if not self.account_key_pem:
            raise RuntimeError("ACME account key was not initialized")
        resp = await self._post(
            self.directory["newAccount"], {"termsOfServiceAgreed": True}
        )
        self.account_url = resp.headers.get("Location")
        await db.save_acme_account(domain, self.account_key_pem, self.account_url)

    async def create_order(self, domains: List[str]) -> dict:
        ids = [{"type": "dns", "value": d} for d in domains]
        resp = await self._post(self.directory["newOrder"], {"identifiers": ids})
        order = resp.json()
        order["_url"] = resp.headers.get("Location")
        return order

    async def get_authorization(self, url: str) -> dict:
        resp = await self._post(url, None)
        return resp.json()

    async def respond_challenge(self, url: str):
        await self._post(url, {})

    async def poll_status(self, url: str, target: str, max_tries: int = 30) -> dict:
        for _ in range(max_tries):
            resp = await self._post(url, None)
            obj = resp.json()
            status = obj.get("status")
            if status == target:
                return obj
            if status in ("invalid", "expired", "revoked"):
                raise Exception(f"状态异常: {status} - {json.dumps(obj, ensure_ascii=False)}")
            await asyncio.sleep(3)
        raise Exception("轮询超时")

    async def finalize(self, url: str, csr_der: bytes) -> dict:
        resp = await self._post(url, {"csr": _b64url(csr_der)})
        return resp.json()

    async def download_cert(self, url: str) -> str:
        resp = await self._post(url, None)
        return resp.text

    # ── 辅助方法 ───────────────────────────────────────────────

    def dns_txt_value(self, token: str) -> str:
        """计算 DNS-01 验证所需的 TXT 记录值"""
        ka = f"{token}.{self._thumbprint()}"
        return _b64url(hashlib.sha256(ka.encode()).digest())

    @staticmethod
    def generate_csr(domains: List[str]) -> Tuple[bytes, bytes]:
        """生成 CSR 和私钥，返回 (csr_der, key_pem)"""
        key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
        name = x509.Name([x509.NameAttribute(NameOID.COMMON_NAME, domains[0])])
        san = x509.SubjectAlternativeName([x509.DNSName(d) for d in domains])
        csr = (
            x509.CertificateSigningRequestBuilder()
            .subject_name(name)
            .add_extension(san, critical=False)
            .sign(key, hashes.SHA256())
        )
        csr_der = csr.public_bytes(serialization.Encoding.DER)
        key_pem = key.private_bytes(
            serialization.Encoding.PEM,
            serialization.PrivateFormat.PKCS8,
            serialization.NoEncryption(),
        )
        return csr_der, key_pem

    # ── 完整申请流程 ───────────────────────────────────────────

    async def apply_certificate(
        self,
        domain: str,
        cf: CloudflareService,
        staging: bool = False,
    ) -> dict[str, Any]:
        """完整的通配符证书申请流程"""
        domains = [domain, f"*.{domain}"]
        dns_records: list[tuple[str, str]] = []

        try:
            await self._log(f"开始申请证书: {', '.join(domains)}", step="init")

            # 1. 初始化
            await self._log("连接 Let's Encrypt...", step="init")
            await self.init(domain)
            await self._log("连接成功", "success", step="init")

            # 2. 注册账户
            await self._log("准备 ACME 账户...", step="account")
            await self.ensure_account(domain)
            await self._log("账户就绪", "success", step="account")

            # 3. 创建订单
            await self._log("创建证书订单...", step="order")
            order = await self.create_order(domains)
            await self._log(f"订单已创建，需验证 {len(order['authorizations'])} 个授权", "success", step="order")

            # 4. 获取 Zone ID
            await self._log(f"查询 Cloudflare Zone: {domain}", step="dns")
            zone_id = await cf.get_zone_id(domain)
            await self._log("Zone 已找到", "success", step="dns")

            # 5. 处理每个授权的 DNS-01 Challenge
            challenges_to_respond = []
            for auth_url in order["authorizations"]:
                auth = await self.get_authorization(auth_url)
                auth_domain = auth["identifier"]["value"]

                dns_ch = next((c for c in auth["challenges"] if c["type"] == "dns-01"), None)
                if not dns_ch:
                    raise Exception(f"未找到 {auth_domain} 的 dns-01 验证")

                txt_val = self.dns_txt_value(dns_ch["token"])
                record_name = f"_acme-challenge.{auth_domain}"

                await self._log(f"添加 DNS 记录: {record_name}", step="dns")
                rid = await cf.create_txt_record(zone_id, record_name, txt_val)
                dns_records.append((zone_id, rid))
                await self._log("DNS 记录已添加", "success", step="dns")

                challenges_to_respond.append((auth_url, dns_ch["url"], auth_domain))

            # 6. 等待 DNS 传播
            await self._log("等待 DNS 传播...", step="propagation")
            for i in range(30, 0, -1):
                await self._log(f"倒计时 {i}s...", "debug", step="propagation")
                await asyncio.sleep(1)
            await self._log("DNS 传播完成", "success", step="propagation")

            # 7. 响应 Challenge
            for auth_url, ch_url, auth_domain in challenges_to_respond:
                await self._log(f"提交验证: {auth_domain}", step="verify")
                await self.respond_challenge(ch_url)

            # 8. 等待验证通过
            for auth_url, _, auth_domain in challenges_to_respond:
                await self._log(f"等待验证: {auth_domain}...", step="verify")
                await self.poll_status(auth_url, "valid")
                await self._log(f"{auth_domain} 验证通过", "success", step="verify")

            # 9. 生成 CSR & 完成订单
            await self._log("生成证书密钥和 CSR...", step="generate")
            csr_der, key_pem = self.generate_csr(domains)
            await self._log("CSR 已生成", "success", step="generate")

            await self._log("提交订单完成请求...", step="finalize")
            result = await self.finalize(order["finalize"], csr_der)

            if result.get("status") != "valid":
                result = await self.poll_status(order["_url"], "valid")

            # 10. 下载证书
            await self._log("下载证书...", step="finalize")
            cert_pem = await self.download_cert(result["certificate"])
            await self._log("证书已下载", "success", step="finalize")

            # 11. 解析证书元数据，调用方负责持久化到 R2/D1
            cert_obj = x509.load_pem_x509_certificate(cert_pem.encode())
            expires = cert_obj.not_valid_after_utc
            issued_at = datetime.now(timezone.utc).isoformat()
            meta = {
                "domain": domain,
                "domains": domains,
                "issued_at": issued_at,
                "expires_at": expires.isoformat(),
            }

            await self._log(f"证书申请成功！有效期至 {expires.strftime('%Y-%m-%d')}", "success", step="complete")
            return {
                "fullchain_pem": cert_pem,
                "privkey_pem": key_pem.decode("utf-8"),
                "metadata": meta,
                "issued_at": issued_at,
                "expires_at": expires.isoformat(),
            }

        except Exception as e:
            await self._log(f"申请失败: {e}", "error", step="error")
            raise
        finally:
            # 清理 DNS 记录
            if dns_records:
                await self._log("清理 DNS 记录...", step="cleanup")
                for zid, rid in dns_records:
                    try:
                        await cf.delete_txt_record(zid, rid)
                    except Exception as e:
                        await self._log(f"清理失败: {e}", "warn", step="cleanup")
                await self._log("DNS 记录已清理", "success", step="cleanup")
            await self.client.aclose()
